File: studio/ratelimit.py
"""Sliding-window rate limiter for auth endpoints."""

# ponytail: in-memory, single-process only; move counters to Mongo if the
# studio backend ever runs multiple instances.

import time

_buckets = {}


def _prune(key, window_s):
    now = time.time()
    bucket = _buckets.setdefault(key, [])
    bucket[:] = [ts for ts in bucket if now - ts < window_s]
    return bucket


def allow(key, limit, window_s):
    bucket = _prune(key, window_s)
    if len(bucket) >= limit:
        return False
    bucket.append(time.time())
    return True


def retry_after(key, window_s):
    bucket = _prune(key, window_s)
    if not bucket:
        return 0
    return max(1, int(window_s - (time.time() - bucket[0])))


def reset_for_tests():
    _buckets.clear()

File: src/server/rateLimit.ts
/** Sliding-window rate limiter (single-process; mirrors studio/ratelimit.py). */

const buckets = new Map<string, number[]>();

function prune(key: string, windowS: number): number[] {
  const now = Date.now() / 1000;
  const bucket = buckets.get(key) ?? [];
  const fresh = bucket.filter((ts) => now - ts < windowS);
  buckets.set(key, fresh);
  return fresh;
}

export function allow(key: string, limit: number, windowS: number): boolean {
  const bucket = prune(key, windowS);
  if (bucket.length >= limit) return false;
  bucket.push(Date.now() / 1000);
  buckets.set(key, bucket);
  return true;
}

export function retryAfter(key: string, windowS: number): number {
  const bucket = prune(key, windowS);
  if (!bucket.length) return 0;
  return Math.max(1, Math.floor(windowS - (Date.now() / 1000 - bucket[0]!)));
}

export function resetForTests(): void {
  buckets.clear();
}
